### PR TITLE
Reject multiple data paths on frozen capable nodes

### PR DESCRIPTION
**Pull request by henningandersen**
Nodes with a frozen cache no longer supports multiple data paths. This
simplifies cache sizing and avoids the need to support multiple cache
files.

Relates #71844 (which also contains the single line documentation on this).

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/shared/FrozenCacheService.java [hunk 1]
```diff
@@ -117,12 +118,23 @@ public void validate(final ByteSizeValue value, final Map<Setting<?>, Object> se
                             roles.stream().map(DiscoveryNodeRole::roleName).collect(Collectors.joining(","))
                         );
                     }
+
+                    @SuppressWarnings("unchecked")
+                    final List<String> dataPaths = (List<String>) settings.get(Environment.PATH_DATA_SETTING);
+                    if (dataPaths.size() > 1) {
+                        throw new SettingsException(
+                            "setting [{}={}] is not permitted on nodes with multiple data paths [{}]",
```
Comment (ywelsch) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/shared/FrozenCacheService.java:126]: How does this interact with the other PR (default frozen cache size) when you have multiple data paths defined and a dedicated frozen node defined, but have not explicitly configured a cache size?

**Review by ywelsch — COMMENTED** [review]
i've left just one comment, o.w. looking good.

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/BaseSearchableSnapshotsIntegTestCase.java [hunk 2]
```diff
@@ -185,4 +202,111 @@ protected void populateIndex(String indexName, int maxIndexRequests) throws Inte
...
+                    final long translogFiles = dir.filter(path -> path.getFileName().toString().contains("translog")).count();
+                    if (snapshotDirectory) {
+                        assertEquals(2L, translogFiles);
+                    } else {
+                        assertThat(translogFiles, greaterThanOrEqualTo(2L));
+                    }
+                }
+            }
+        }
+        assertTrue("no shard folder found", shardFolderFound);
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/BaseSearchableSnapshotsIntegTestCase.java:247]: Maybe indicate the name of the index in the error?

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/BaseSearchableSnapshotsIntegTestCase.java [hunk 3]
```diff
@@ -185,4 +202,111 @@ protected void populateIndex(String indexName, int maxIndexRequests) throws Inte
...
+        final Index restoredIndex = resolveIndex(indexName);
+        final String customDataPath = resolveCustomDataPath(indexName);
+        final ShardId shardId = new ShardId(restoredIndex, 0);
+        boolean shardFolderFound = false;
+        for (String node : internalCluster().getNodeNames()) {
+            final NodeEnvironment service = internalCluster().getInstance(NodeEnvironment.class, node);
+            final ShardPath shardPath = ShardPath.loadShardPath(logger, service, shardId, customDataPath);
+            if (shardPath != null && Files.exists(shardPath.getDataPath())) {
+                shardFolderFound = true;
+                assertEquals(snapshotDirectory, Files.notExists(shardPath.resolveIndex()));
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/BaseSearchableSnapshotsIntegTestCase.java:234]: I guess we could determine `snapshotDirectory` from the index settings?

**Review by tlrx — APPROVED** [review]
LGTM

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/shared/FrozenCacheService.java (hunk 1 above)
Comment (henningandersen) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/shared/FrozenCacheService.java:126]: It is my intention to make that case fail once I merge the two together.

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/BaseSearchableSnapshotsIntegTestCase.java (hunk 2 above)
Comment (henningandersen) [reply x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/BaseSearchableSnapshotsIntegTestCase.java]: Added in https://github.com/elastic/elasticsearch/pull/71896#discussion_r616494070

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/BaseSearchableSnapshotsIntegTestCase.java (hunk 3 above)
Comment (henningandersen) [reply x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/BaseSearchableSnapshotsIntegTestCase.java:234]: I simply moved this method from `SearchableSnapshotsIntegTests` so will defer improving this to a follow-up.

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**